### PR TITLE
port partr multiq to julia

### PR DESCRIPTION
**Pull request by vtjnash**
Direct translation, not necessarily fully idiomatic. In preparation for
future improvements.

### CODE REVIEW

File: base/partr.jl [hunk 1]
```diff
@@ -0,0 +1,164 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module Partr
+
+# a task heap
+mutable struct taskheap
+    const lock::ReentrantLock
```
Comment (tkf) [inline base/partr.jl:7]: We only use `trylock` so I guess it's fine, but isn't it a bit weird to use a lock that requires the scheduler for implementing the scheduler? Actually, why not just use a single atomic bool? I think it's fine as long as we have more heaps than the number of threads.

File: src/partr.c [hunk 2]
```diff
@@ -299,18 +131,6 @@ void jl_threadfun(void *arg)
...
 
 
-// enqueue the specified task for execution
-JL_DLLEXPORT int jl_enqueue_task(jl_task_t *task)
-{
-    char failed;
-    if (multiq_insert(task, task->prio) == -1)
-        failed = 1;
-    failed = 0;
-    JL_PROBE_RT_TASKQ_INSERT(jl_current_task->ptls, task);
```
Comment (vchuravy) [inline src/partr.c:309]: Losing this is disappointing :) We will need to find a way to implement this from Julia efficiently.

Comment (jpsamaroo) [reply src/partr.c:309]: Could we implement the logic of https://github.com/JuliaPerf/UProbes.jl directly in codegen?

Comment (vchuravy) [reply src/partr.c:309]: We could, but bpftrace would still not find the new probes. UProbes.jl works by outlining the probe to a shared library that bpftrace and co. can scan.

File: base/partr.jl (hunk 1 above)
Comment (vtjnash) [reply base/partr.jl:7]: Good point. I would actually like to make this an `InvasiveLinkedListSynchronized{Task}` (the same type as `StickyWorkqueue`). But for the initial port, `Threads.SpinLock` would be a more direct translation.

**Comment by kpamnany** [comment]
What was the performance impact of moving this?

**Comment by vtjnash** [comment]
What would you recommend for a benchmark?

**Comment by kpamnany** [comment]
I have something for parallel queues somewhere. Will find and post.

**Comment by vtjnash** [comment]
After working around https://github.com/JuliaLang/julia/issues/44720, I see similar timing, but 5-10x less allocation with this PR

**Pull request merged**